loki/transform/fortran_c_transform.py

```diff
@@ -477,17 +478,51 @@ def generate_c_kernel(self, routine):
         convert_to_lower_case(kernel)
 
         # Force pointer on reference-passed arguments
-        var_map = {}
+        to_be_dereferenced = []
         for arg in kernel.arguments:
             if not(arg.type.intent.lower() == 'in' and isinstance(arg, Scalar)):
                 _type = arg.type.clone(pointer=True)
                 if isinstance(arg.type.dtype, DerivedType):
                     # Lower case type names for derived types
                     typedef = _type.dtype.typedef.clone(name=_type.dtype.typedef.name.lower())
                     _type = _type.clone(dtype=typedef.dtype)
-                var_map[arg] = Dereference(arg)
+                to_be_dereferenced.append(arg.name.lower())
                 kernel.symbol_attrs[arg.name] = _type
-        kernel.body = SubstituteExpressions(var_map).visit(kernel.body)
+
+        class DeReferenceTrafo(Transformer):
```

Comment:
The criterion for a symbol that needs (de)referencing is the same throughout, so maybe it would make sense to capture this as a static method in the transformer:

```py
@staticmethod
def is_reference(symbol):
    return (
        isinstance(symbol, Array) 
        and not (symbol.dimensions is None or all(dim == ':' for dim in symbol.dimensions))
    )
```